File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/storage/NitroPlayerStorage.kt
package com.margelo.nitro.nitroplayer.storage

import android.content.Context
import com.margelo.nitro.nitroplayer.core.NitroPlayerLogger
import java.io.File

object NitroPlayerStorage {
    private const val TAG = "NitroPlayerStorage"
    private const val DIR_NAME = "nitroplayer"

    /** Reads the contents of [filename] from the NitroPlayer storage directory, or null if absent. */
    fun read(context: Context, filename: String): String? {
        val file = File(storageDirectory(context), filename)
        return if (file.exists()) {
            try {
                file.readText()
            } catch (e: Exception) {
                NitroPlayerLogger.log(TAG, "read($filename) failed: $e")
                null
            }
        } else {
            null
        }
    }

    /**
     * Atomically writes [json] to [filename] in the NitroPlayer storage directory.
     * Writes to `<filename>.tmp` first, then renames — leaving the prior file
     * untouched on failure (crash-safe).
     */
    fun write(context: Context, filename: String, json: String) {
        try {
            val dir = storageDirectory(context)
            dir.mkdirs()
            val tmp = File(dir, "$filename.tmp")
            val dest = File(dir, filename)
            tmp.writeText(json)
            if (!tmp.renameTo(dest)) {
                // renameTo can fail across mount points; copy then delete as fallback
                dest.writeText(tmp.readText())
                tmp.delete()
            }
        } catch (e: Exception) {
            NitroPlayerLogger.log(TAG, "write($filename) failed: $e")
        }
    }

    /** Returns the NitroPlayer subdirectory inside filesDir. */
    private fun storageDirectory(context: Context): File = File(context.filesDir, DIR_NAME)
}

File: react-native-nitro-player/ios/storage/NitroPlayerStorage.swift
//
//  NitroPlayerStorage.swift
//  NitroPlayer
//
//  Created by Ritesh Shukla on 19/02/26.
//

import Foundation

enum NitroPlayerStorage {
  /// Reads raw data from a file in the NitroPlayer storage directory.
  /// Returns nil if the file does not exist or cannot be read.
  static func read(filename: String) -> Data? {
    let url = storageDirectory().appendingPathComponent(filename)
    return try? Data(contentsOf: url)
  }

  /// Atomically writes data to a file in the NitroPlayer storage directory.
  /// Writes to `<filename>.tmp` first, then renames to the final name —
  /// leaving the prior file untouched if the write crashes mid-way.
  static func write(filename: String, data: Data) throws {
    let dir = storageDirectory()
    try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    let dest = dir.appendingPathComponent(filename)
    let tmp = dir.appendingPathComponent(filename + ".tmp")
    try data.write(to: tmp)
    if FileManager.default.fileExists(atPath: dest.path) {
      _ = try FileManager.default.replaceItemAt(dest, withItemAt: tmp)
    } else {
      try FileManager.default.moveItem(at: tmp, to: dest)
    }
  }

  /// Returns the NitroPlayer subdirectory inside Application Support.
  /// Uses `FileManager` APIs — never hardcodes the UUID-based container path
  /// so this resolves correctly regardless of which device or simulator the
  /// app runs on.
  private static let storageDir: URL = {
    let appSupport = FileManager.default.urls(
      for: .applicationSupportDirectory, in: .userDomainMask
    ).first!
    return appSupport.appendingPathComponent("NitroPlayer", isDirectory: true)
  }()

  private static func storageDirectory() -> URL { storageDir }
}
